test/e2e/apps/statefulset.go

```diff
@@ -1339,6 +1342,82 @@ var _ = SIGDescribe("StatefulSet", func() {
 			framework.ExpectNoError(err)
 		})
 
+		ginkgo.It("should not delete PVC with OnScaledown policy if another controller owns the PVC", func(ctx context.Context) {
+			e2epv.SkipIfNoDefaultStorageClass(ctx, c)
+			ginkgo.By("Creating statefulset " + ssName + " in namespace " + ns)
+			*(ss.Spec.Replicas) = 3
+			_, err := c.AppsV1().StatefulSets(ns).Create(ctx, ss, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Confirm PVC has been created")
+			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{0, 1, 2})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Create configmap to use as dummy controller")
+			dummyConfigMap, err := c.CoreV1().ConfigMaps(ns).Create(ctx, &v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "dummy-controller",
+					Namespace:    ns,
+				},
+			}, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+			defer func() {
+				// Will be cleaned up with the namespace if this fails.
+				_ = c.CoreV1().ConfigMaps(ns).Delete(ctx, dummyConfigMap.Name, metav1.DeleteOptions{})
+			}()
+
+			ginkgo.By("Update PVC 1 owner ref")
+			pvc, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, fmt.Sprintf("datadir-%s-1", ss.Name), metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			pvc.OwnerReferences = []metav1.OwnerReference{
+				{
+					Name:       dummyConfigMap.GetName(),
+					APIVersion: "v1",
+					Kind:       "ConfigMap",
+					UID:        dummyConfigMap.GetUID(),
+					Controller: ptr.To(true),
+				},
+			}
+			pvc, err = c.CoreV1().PersistentVolumeClaims(ns).Update(ctx, pvc, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Update StatefulSet retention policy")
+			ss.Spec.PersistentVolumeClaimRetentionPolicy = &appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+				WhenScaled: appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
+			}
+			ss, err = c.AppsV1().StatefulSets(ns).Update(ctx, ss, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Scale StatefulSet down to 0")
+			_, err = e2estatefulset.Scale(ctx, c, ss, 0)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Verify PVC 1 still exists")
+			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{1})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Remove PVC 1 owner ref")
+			pvc.OwnerReferences = nil
+			_, err = c.CoreV1().PersistentVolumeClaims(ns).Update(ctx, pvc, metav1.UpdateOptions{})
```

Comment:
I've tried running this test locally and the first two invocations I hit:
```
  [FAILED] Operation cannot be fulfilled on persistentvolumeclaims "datadir-ss-1": the object has been modified; please apply your changes to the latest version and try again
  In [It] at: k8s.io/kubernetes/test/e2e/apps/statefulset.go:1402 @ 06/04/24 12:13:52.29
```
so I'd suggest to go through all the PVCs update operations and make them retriable. Similarly StatefulSet updates should be changed to use `updateStatefrulSetWithRetries` most likely.

Reply:
Done.

Interesting, I've been using kind and never hit that. But +1 to fewer flakes.